admin/server/postgres.go

```diff
@@ -0,0 +1,204 @@
+package server
+
+import (
+	"context"
+	"fmt"
+	"net/url"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/jackc/pgx/v5/pgxpool"
+	wire "github.com/jeroenrinzema/psql-wire"
+	"github.com/lib/pq/oid"
+	"github.com/rilldata/rill/admin/server/auth"
+	runtimeauth "github.com/rilldata/rill/runtime/server/auth"
+	"go.uber.org/zap"
+)
+
+func (s *Server) QueryHandler(ctx context.Context, query string) (wire.PreparedStatements, error) {
+	s.logger.Debug("query", zap.String("query", query))
+	if strings.Trim(query, " ") == "" {
+		return wire.Prepared(wire.NewStatement(func(ctx context.Context, writer wire.DataWriter, parameters []wire.Parameter) error {
+			return writer.Empty()
+		})), nil
+	}
+
+	upperQuery := strings.ToUpper(query)
+	if strings.HasPrefix(upperQuery, "SET") {
+		return wire.Prepared(wire.NewStatement(func(ctx context.Context, writer wire.DataWriter, parameters []wire.Parameter) error {
+			return writer.Complete("SET")
+		}, wire.WithColumns(nil))), nil
+	}
+
+	if strings.HasPrefix(upperQuery, "BEGIN") || strings.HasPrefix(upperQuery, "COMMIT") || strings.HasPrefix(upperQuery, "ROLLBACK") {
+		return wire.Prepared(wire.NewStatement(func(ctx context.Context, writer wire.DataWriter, parameters []wire.Parameter) error {
+			return writer.Complete(strings.Trim(upperQuery, ";"))
+		}, wire.WithColumns(nil))), nil
+	}
+
+	clientParams := wire.ClientParameters(ctx)
+	// database is org.password
+	tokens := strings.Split(clientParams[wire.ParamDatabase], ".")
+	if len(tokens) != 2 {
+		return nil, fmt.Errorf("invalid org or project")
+	}
+	org := tokens[0]
+	project := tokens[1]
+
+	// Find the production deployment for the project we're proxying to
+	proj, err := s.admin.DB.FindProjectByName(ctx, org, project)
+	if err != nil {
+		return nil, fmt.Errorf("invalid org or project")
+	}
+
+	if proj.ProdDeploymentID == nil {
+		return nil, fmt.Errorf("no prod deployment for project")
+	}
+	depl, err := s.admin.DB.FindDeployment(ctx, *proj.ProdDeploymentID)
+	if err != nil {
+		return nil, fmt.Errorf("no prod deployment for project")
+	}
+
+	var jwt string
+	claims := auth.GetClaims(ctx)
+	switch claims.OwnerType() {
+	case auth.OwnerTypeAnon:
+		// If the client is not authenticated with the admin service, we just proxy the contents of the password to the runtime (if any).
+		password := ctx.Value(auth.PostgresPassword{}).(string)
+		if len(password) >= 6 && strings.EqualFold(password[0:6], "bearer") {
+			jwt = strings.TrimSpace(password[6:])
+		}
```

Comment:
I saw `wire.SessionHandler` previously but didn't understood it fully. I looked into the code again and yeah seems like we can set JWT. To avoid reconnecting I had setup pgxpool cache but yeah with the `SessionHandler` we can avoid token validation calls as well. 
I guess we would still need the pgxpool cache to handle closing the connections on server close.